### DISCUSSION TITLE
Prefetching / caching pages objects

### DISCUSSION THREAD

**Discussion by buhrmi** [inertiajs/inertia#766] [Ideas]
Some frameworks offer prefetching and caching of routes to make the application feel snappier. For example, Sapper implements the `prefetch` attribute on `a` tags, which will cause the route to be preloaded as soon as the user hovers over the link, and instantly replaces the content when the user clicks it. 

Other frameworks, like Turbolinks, cache whole routes, so that the user is briefly presented with stale content after navigation, while the "fresh" content is being loaded in the background.

My proposal here is for Inertia to implement something similar via prefetching and caching page objects. 

Would like to hear thoughts.

**Comment by jonagoldman** [comment]
I think this is out of the scope of the library. You can achieve that with dynamic imports or using a bundler like Vite that already has it.

**Reply by buhrmi** [reply]
No you can not

**Reply by jonagoldman** [reply]
Alright....

> [SSG Pre-Rendering ](https://vitejs.dev/guide/ssr.html#pre-rendering-ssg)
>  If the routes and the data needed for certain routes are known ahead of time, we can pre-render these routes into static HTML using the same logic as production SSR.

> [Lazy Loaded Dynamic Imports](https://vitejs.dev/guide/features.html#glob-import)
> Matched files are by default lazy loaded via dynamic import and split into separate chunks. If you'd rather import all the modules directly you can use eager loading instead.

> [Link prefetching](https://developer.mozilla.org/en-US/docs/Glossary/Prefetch#link_prefetching)
> Link prefetching is a performance optimization technique that works by assuming which links the user is likely to click, then downloading the content of those links. If the user decides to click on one of the links, then the page will be rendered instantly as the content has already been downloaded.

> [ES Native Dynamic Imports](https://hospodarets.com/native-ecmascript-modules-dynamic-import)
> [ES6 Dynamic Import](https://caniuse.com/es6-module-dynamic-import)
> [SEC Import Calls](https://tc39.es/ecma262/#sec-import-calls) 

Cheers.

**Reply by buhrmi** [reply]
dude this has nothing to do with prefetching / caching the inertia page object

**Reply by ncphillips** [reply]
Big +1 for @buhrmi here. 

None of those links solve this problem.

* `SSG Pre-Rendering` is about pre-rendering HTML. It would only work for first loading a page and wouldn't effect client-side navigation at all.
* `Lazy Loaded Dynamic Imports`, `ES Native Dynamic Imports`, `ES6 Dynamic Import`,  and `SEC Import Calls`  is completely unrelated. That's loading JS files, not the JSON returned from the page's endpoint.
* `Link Prefetching` is the closest thing to an actual answer...but it doesn't work with Inertia. The Inertia link/router architecture hijacks normal link behaviour.

**Comment by reinink** [comment]
Yeah, would be cool. This is something @claudiodekker has spent some time on (see #223), but it's not something we've had time to finish. Maybe in the future. 👍

**Reply by buhrmi** [reply]
Ah, nice. Yeah #223 is exactly what I meant. That would be really great for clients who want true instant switching between different pages on mobile apps. Even it means to briefly see stale content.

**Reply by jonagoldman** [reply]
Oh, I see now, totally misunderstood the request. My apologies. 🙏

**Reply by buhrmi** [reply]
Now I think that the functionality for prefetching & caching page objects should live inside a service worker 🤔

**Reply by claudiodekker** [reply]
Yeah, we've considered that, but Service Workers are another step entirely, and require changes on a larger scale.
For now, if we're going to implement it at some point in the future, it'll most likely be as in that #223 PR example.

**Comment by joydeep-bhowmik** [comment]
Wish it was implemented

**Comment by ncphillips** [comment]
Definitely wish this was a feature. 

Without preloading, Inertia definitely feels much less snappy than other frameworks I've worked with. [Next.js](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#2-prefetching) in particular feels great.

While there is a native prefetch ability in browsers, it doesn't seem to work with Inertia. I assume this is because we're loading JSON.

**Reply by buhrmi** [reply]
Add it to the inertia 2.0 roadmap 😆😛
